test/distributed/_composable/test_fully_shard.py

```diff
@@ -452,44 +455,148 @@ def world_size(self) -> int:
         local_model = CompositeParamModel(device=torch.device("cuda"))
-        composable_module = copy.deepcopy(local_model)
-        fully_shard(composable_module, policy=ModuleWrapPolicy({UnitModule}))
+        save_composable = copy.deepcopy(local_model)
+        fully_shard(save_composable, policy=ModuleWrapPolicy({UnitModule}))
         local_sd = local_model.state_dict()
-        composable_sd = composable_module.state_dict()
+        composable_sd = save_composable.state_dict()
         self._check_state_dict_parity(local_sd, composable_sd)
 
+        # Validate load
+        load_composable = fully_shard(
+            copy.deepcopy(local_model),
+            policy=ModuleWrapPolicy({UnitModule})
+        )
+        _zero_model(load_composable)
+        for p in load_composable.parameters():
+            self.assertEqual(p.sum(), 0)
+
+        sd = {k: v.clone() for k, v in composable_sd.items()}
+        print(f"sd keys: {list(sd.keys())}")
+        load_composable.load_state_dict(sd)
+        self._check_model_parity(load_composable, save_composable)
+
     @skip_if_lt_x_gpu(2)
-    def test_state_dict_submodule_fully_shard(self):
+    def test_state_dict_save_load_submodule_fully_shard(self):
         """
         Tests that the full state dict saved from a module with ``fully_shard``
-        applied on submodules matches that of an equivalent local module.
+        applied on submodules matches that of an equivalent local module. Also
+        ensures that this state_dict can be reloaded into a composable module and
+        is equivalent to the original composable module.
         """
         local_model = CompositeParamModel(device=torch.device("cuda"))
-        composable_module = copy.deepcopy(local_model)
-        fully_shard(composable_module.u1)
-        fully_shard(composable_module.u2)
+
+        def _create_fully_shard_on_submodules(mod: nn.Module):
+            fully_shard(mod.u1)
+            fully_shard(mod.u2)
+            return mod
+
+        save_composable = copy.deepcopy(local_model)
+        save_composable = _create_fully_shard_on_submodules(save_composable)
         local_sd = local_model.state_dict()
-        composable_sd = composable_module.state_dict()
+        composable_sd = save_composable.state_dict()
         self._check_state_dict_parity(local_sd, composable_sd)
 
+        # Validate load
+        load_composable = copy.deepcopy(local_model)
+        load_composable = _create_fully_shard_on_submodules(load_composable)
+        _zero_model(load_composable)
+        for p in load_composable.parameters():
+            self.assertEqual(0, p.sum())
+
+        sd = {k: v.clone() for k, v in composable_sd.items()}
+        load_composable.load_state_dict(sd)
+        self._check_model_parity(load_composable, save_composable)
+
+    @skip_if_lt_x_gpu(2)
+    def test_state_dict_save_load_flow(self):
+        """
+        E2E test of save + load with rank0_only + CPU offload for TransformerWithSharedParams
```

Comment:
In the future, will this test include different state dict types and subtest the different configs?